### ISSUE TITLE
Causal lanuage modeling should consider loss masks for pad tokens.

### ISSUE THREAD

**Issue by sherlcok314159** [huggingface/transformers#32832]
### Feature request

A common snippet for training causal language models is like the below:

```python
        loss = None
        if labels is not None:
            # Shift so that tokens < n predict n
            shift_logits = logits[..., :-1, :].contiguous()
            shift_labels = labels[..., 1:].contiguous()
            # Flatten the tokens
            loss_fct = CrossEntropyLoss()
            shift_logits = shift_logits.view(-1, self.config.vocab_size)
            shift_labels = shift_labels.view(-1)
            # Enable model parallelism
            shift_labels = shift_labels.to(shift_logits.device)
            loss = loss_fct(shift_logits, shift_labels)

        if not return_dict:
            output = (logits,) + outputs[1:]
            return (loss,) + output if loss is not None else output
```

However, in real-world scenarios, we train LLMs by batched inputs which results in pad tokens in batches. Those pad tokens should be neglected for loss computation. For instance, switch pad token labels to `-100`.

### Motivation

See the above

### Your contribution

If you are busy for this, I can submit a PR.

**Comment by amyeroberts** [comment]
Hi @sherlcok314159, please check out our NLP course and how to use data collators to handle setting padding to -100: https://huggingface.co/learn/nlp-course/chapter7/4?fw=pt#data-collation

**Comment by sherlcok314159** [comment]
Thanks for your kind advice. I found Trainer already has the collator function to ignore pad tokens.
